servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/SingleConcatWithPublisher.java

```diff
@@ -43,100 +47,141 @@ Executor executor() {
-         *     <li>Actual result if {@link #onSuccess(Object)} invoked before {@link #request(long)}</li>
-         *     <li>{@link ConcatSubscriber#REQUESTED} if {@link #request(long)} (with a valid n) invoked before
-         *     {@link #onSuccess(Object)}</li>
-         *     <li>{@link ConcatSubscriber#CANCELLED} if {@link #cancel()} is called or the first call to request(n)
-         *     is invalid </li>
-         * </ul>
+         * It may be the actual result if {@link #onSuccess(Object)} invoked before {@link #request(long)}.
          */
         @Nullable
-        private volatile Object mayBeResult = INITIAL;
+        volatile Object mayBeResult = INITIAL;
 
-        ConcatSubscriber(final Subscriber<? super T> subscriber, final Publisher<? extends T> next) {
-            this.target = subscriber;
+        AbstractConcatSubscriber(final Subscriber<? super T> target, final Publisher<? extends T> next) {
+            this.target = target;
             this.next = next;
         }
 
         @Override
-        public void onSubscribe(final Cancellable cancellable) {
+        public final void onSubscribe(final Cancellable cancellable) {
             delayedCancellable(cancellable);
             target.onSubscribe(this);
         }
 
         @Override
-        public void onSuccess(@Nullable final T result) {
-            for (;;) {
-                Object oldValue = mayBeResult;
-                if (oldValue == REQUESTED) {
-                    emitSingleSuccessToTarget(result);
-                    break;
-                } else if (oldValue == CANCELLED || mayBeResultUpdater.compareAndSet(this, INITIAL, result)) {
-                    break;
-                }
-            }
+        public final void onSubscribe(final Subscription subscription) {
+            delayedSubscription(subscription);
         }
 
         @Override
-        public void onComplete() {
-            target.onComplete();
+        public final void onNext(@Nullable final T t) {
+            target.onNext(t);
         }
 
         @Override
-        public void onSubscribe(final Subscription subscription) {
-            delayedSubscription(subscription);
+        public final void onError(final Throwable t) {
+            target.onError(t);
         }
 
         @Override
-        public void onNext(@Nullable final T t) {
-            target.onNext(t);
+        public final void onComplete() {
+            target.onComplete();
         }
 
         @Override
-        public void onError(final Throwable t) {
-            target.onError(t);
+        public final void cancel() {
+            // We track cancelled here because we need to make sure if cancel() happens subsequent calls to request(n)
+            // are NOOPs [1].
+            // [1] https://github.com/reactive-streams/reactive-streams-jvm#3.6
+            mayBeResult = CANCELLED;
+            super.cancel();
+        }
+
+        final void superCancel() {
```

Comment:
this seems a little odd, what problem is it trying to solve?

Reply:
haha, the name of the method is funny :) 
the intend is to invoke `DelayedCancellableThenSubscription#cancel()` from subclasses of `AbstractConcatSubscriber`. Kind of `super.super.cancel()`. I will add a comment